### PR TITLE
Fix internal slot in event handlers

### LINKED ISSUE #447: Check Internal slots for events
_Originally posted by @relu91 in https://github.com/w3c/wot-scripting-api/issues/424#issuecomment-1356332947_

### PR DESCRIPTION
**Pull request by relu91**
Should eventually fix #447.

***
<a href="https://pr-preview.s3.amazonaws.com/relu91/wot-scripting-api/pull/463.html" title="Last updated on Mar 20, 2023, 5:22 PM UTC (33861b7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-scripting-api/463/f950ea5...relu91:33861b7.html" title="Last updated on Mar 20, 2023, 5:22 PM UTC (33861b7)">Diff</a>

### CODE REVIEW

**Comment by zolkis** [comment]
Should we use https://webidl.spec.whatwg.org/#record-type instead of maps?

**Comment by zolkis** [comment]
Also, instead of pair, we should use tuple: https://infra.spec.whatwg.org/#tuple

**Review by zolkis — APPROVED** [review]

**Comment by zolkis** [comment]
I think this PR completes its original scope. The above comments can be fixed in a new PR that fixes #448 (assigned to me).

I suggest we merge this, so that I can continue from there.

**Comment by danielpeintner** [comment]
Smaller PRs are fine by me 👍.

@zolkis would you take over the remaining points mentioned in this issue? 
e.g., pair to tuple, record-type instead of maps

**Review by danielpeintner — APPROVED** [review]
> I suggest we merge this, so that I can continue from there.

I am okay with merging as is.
However, we should take care of the other aspects in another issue/PR...

**Comment by danielpeintner** [comment]
Follow-up issues created by @zolkis, see https://github.com/w3c/wot-scripting-api/issues/465 and https://github.com/w3c/wot-scripting-api/issues/466

**Comment by relu91** [comment]
Hi all, I think I've completed the work that should be sufficient to fix the linked issue. The text might not be perfect, let me know if you have any ideas to improve it. As @zolkis said other fixes can be stacked on top of this PR, but I would merge it before the next call so that we can move quickly toward a ready-to-be-published spec.

File: index.html [hunk 1]
```diff
@@ -2898,7 +2918,7 @@ <h3>The <dfn>PropertyReadHandler</dfn> callback</h3>
           </li>
           <li>
             Save the request sender information together with |options| to
-            |property|'s <a>internal observer list</a>, in order to be able
+            {{ExposedThing/[[propertyObservers]]}}'s |name| value, in order to be able
```
Comment (zolkis) [inline index.html:2921]: If the above is changed (use a map of sets for the internal slot), the original text is easier to get and more correct.

> [=Append|list-append=] the request sender information together with |options| to |property|'s <a>internal observer list</a>

File: index.html [hunk 2]
```diff
@@ -2605,6 +2605,26 @@ <h4>Internal slots for ExposedThing</h4>
             <td>`{}`</td>
             <td>A {{Map}} with action names as keys and {{ActionHandler}}s as values</td>
           </tr>
+          <tr>
+            <td><dfn>[[\subscribeHandlers]]</dfn></td>
+            <td>`{}`</td>
+            <td>A {{Map}} with event names as keys and {{EventSubscriptionHandler}}s as values</td>
+          </tr>
```
[the remark this answers is not in the document]
Comment (relu91) [reply index.html:2612]: Given your [second comment](https://github.com/w3c/wot-scripting-api/pull/463#discussion_r1142490498), I assume you wanted to comment `[[eventListeners]]` internal slot [change](https://github.com/w3c/wot-scripting-api/pull/463/files#diff-0eb547304658805aad788d320f10bf1f292797b5e6d745a3bf617584da017051R2624), right?

**Comment by danielpeintner** [comment]
> I think we need a second attempt on this, using lists and maps (structures defined in the infra spec).
> https://infra.spec.whatwg.org/

@zolkis Not exactly sure what you mean by that...

File: index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (relu91) [reply index.html:2612]: Sorry, I don't follow. We have two types of internal slots for ExposedThing:
1. slots for handlers -> We need 1 handler for each affordance
2. slots for listeners/observers -> We need a list of listeners for each affordance 

What I don't get is why are you using in both types the world `handlers`.

**Comment by zolkis** [comment]
I explained in this [comment](https://github.com/w3c/wot-scripting-api/pull/463#discussion_r1142486509) above.

We should use structures defined in the INFRA spec, e.g. lists (or sets, or eventually records) as internal slots, as we have standard prose for checking elements etc, and don't need to manually describe array operations.

**Review by danielpeintner — APPROVED** [review]

**Comment by danielpeintner** [comment]
Scripting Call 2023-04-03
* merge this PR as is (introduces consistency)
* follow-up PRs coming to align with the INFRA spec

**Pull request merged**